lib/llhttp/parser.js

```diff
@@ -0,0 +1,240 @@
+    wasm_on_url,
+    wasm_on_status,
+    wasm_on_header_field,
+    wasm_on_header_value,
+    wasm_on_headers_complete,
+    wasm_on_body,
+    wasm_on_message_complete
+  }
+})
+
+inst.exports._initialize() // wasi reactor
+
+class HTTPParserError extends Error {
+  constructor (message, code) {
+    super(message)
+    Error.captureStackTrace(this, HTTPParserError)
+    this.name = 'HTTPParserError'
+    this.code = code ? `HPE_${code}` : undefined
+  }
+}
+
+class HTTPParser {
+  constructor (maxHeaderSize, lenient, headersTimeout) {
+    this[kPtr] = inst.exports.llhttp_alloc(constants.TYPE.RESPONSE)
+    this[kBufferSize] = 0
+    this[kBufferPtr] = null
+    this[kCurrentBuffer] = null
+    this[kStatusMessage] = null
+    this[kHeadersFields] = []
+    this[kHeadersValues] = []
+    this[kMaxHeadersSize] = maxHeaderSize
+    this[kHeadersTimeout] = headersTimeout
+    instMap.set(this[kPtr], this)
+  }
+
+  [kMallocBuffer] (size) {
+    if (this[kBufferPtr]) {
+      inst.exports.free(this[kBufferPtr])
+    }
+    this[kBufferSize] = size
+    this[kBufferPtr] = inst.exports.malloc(size)
+  }
+
+  [kGetHeaders] () {
+    const rawHeaders = []
+    for (let c = 0; c < this[kHeadersFields].length; c++) {
+      rawHeaders.push(this[kHeadersFields][c], this[kHeadersValues][c])
+    }
+    this[kHeadersFields] = []
+    this[kHeadersValues] = []
+    return rawHeaders
+  }
+
+  [kOnMessageBegin] () {
+    return 0
+  }
+
+  [kOnHeaders] (rawHeaders) {}
+
+  [kOnHeadersComplete] (versionMajor, versionMinor, rawHeaders, method,
+    url, statusCode, statusMessage, upgrade, shouldKeepAlive) {
+    return 0
+  }
+
+  [kOnBody] (body) {
+    return 0
+  }
+
+  [kOnMessageComplete] () {
+    return 0
+  }
+
+  [kOnExecute] (ret, buffer) {}
+
+  close () {
+    instMap.delete(this[kPtr])
+    inst.exports.llhttp_free(this[kPtr])
+    this[kPtr] = null
+    inst.exports.llhttp_free(this[kBufferPtr])
+    this[kBufferPtr] = null
+    this[kCurrentBuffer] = Buffer.alloc(0)
+  }
+
+  execute (data) {
+    this[kCurrentBuffer] = data
+    // Be sure the parser buffer can contain `data`
+    if (data.byteLength > this[kBufferSize]) {
+      this[kMallocBuffer](Math.ceil(data.byteLength / 4096) * 4096)
+    }
+    // Instantiate a Unit8 Buffer view of the wasm memory that starts from the parser buffer pointer
+    // and has the size of `data`
+    const u8 = new Uint8Array(inst.exports.memory.buffer, this[kBufferPtr], data.byteLength)
+    // Fill the view with `data`
+    u8.set(data)
+    // Call `execute` on the wasm parser.
+    // We pass the `llhttp_parser` pointer address, the pointer address of buffer view data,
+    // and finally the length of bytes to parse.
+    // The return value is an error code or `constants.ERROR.OK`.
+    // See https://github.com/dnlup/llhttp/blob/undici_wasm/src/native/api.c#L106
+    const err = inst.exports.llhttp_execute(this[kPtr], this[kBufferPtr], data.byteLength)
+    let ret = data.byteLength
```

Comment:
```suggestion
    let ret = data.length
```

Reply:
Asking again because of my ignorance, why is `length` better than `byteLength` in this case?

Reply:
They are equal. My personal preference. Feel free to ignore.